lib/cretonne/meta/isa/intel/encodings.py

```diff
@@ -142,38 +159,52 @@ def enc_flt(inst, recipe, *args, **kwargs):
 I64.enc(base.ctz.i32, *r.urm.rex(0xf3, 0x0f, 0xbc), isap=cfg.use_bmi1)
 I64.enc(base.ctz.i32, *r.urm(0xf3, 0x0f, 0xbc), isap=cfg.use_bmi1)
 
+#
 # Loads and stores.
-I32.enc(base.store.i32.any, *r.st(0x89))
-I32.enc(base.store.i32.any, *r.stDisp8(0x89))
-I32.enc(base.store.i32.any, *r.stDisp32(0x89))
+#
+enc_i32_i64_ld_st(base.store, 1, r.st, 0x89)
+enc_i32_i64_ld_st(base.store, 1, r.stDisp8, 0x89)
+enc_i32_i64_ld_st(base.store, 1, r.stDisp32, 0x89)
+
+I64.enc(base.istore32.i64.any, *r.st.rex(0x89))
+I64.enc(base.istore32.i64.any, *r.stDisp8.rex(0x89))
+I64.enc(base.istore32.i64.any, *r.stDisp32.rex(0x89))
+
+enc_i32_i64_ld_st(base.istore16, 0, r.st, 0x66, 0x89)
+enc_i32_i64_ld_st(base.istore16, 0, r.stDisp8, 0x66, 0x89)
+enc_i32_i64_ld_st(base.istore16, 0, r.stDisp32, 0x66, 0x89)
+
+enc_i32_i64_ld_st(base.istore8, 0, r.st_abcd, 0x88)
+enc_i32_i64_ld_st(base.istore8, 0, r.stDisp8_abcd, 0x88)
+enc_i32_i64_ld_st(base.istore8, 0, r.stDisp32_abcd, 0x88)
```

Comment:
Aha, so this part is extra tricky because you're using 8-bit registers.

In 32-bit mode, and in 64-bit instructions *without* a REX prefix, the following 8-bit registers can be addressed: AL, CL, DL, BL, AH, CH, DH, BH. The last 4 are the "high" byte registers, corresponding to bits 8-15 of the ABCD GPRs.

In 64-bit mode *with* a REX prefix, the accessible 8-bit registers are: AL, CL, DL, BL, SPL, BPL, SIL, DIL, R8B-R15B. Note the difference in registers 4-7 — they are the low bytes of GPRs instead of the high bytes of ABCD.

In Cretonne, we don't model the high-byte registers. They are never used. This is the reason the `I32` byte store instructions have an `ABCD` register constraint instead of `GPR` like the other instructions. When storing a byte, the value to store must be in one of the ABCD registers because without a REX prefix, we can't store the low byte of `%esi` for example.

However, *with* a REX prefix, a byte store can access the low byte of all 16 registers, so we just want a normal `GPR` register constraint which is much easier for the register allocator to satisfy.

The end result is:
- When accessing 8-bit registers *without* a REX prefix, use the `ABCD` register constraint.
- When accessing 8-bit registers *with* a REX prefix, use the `GPR` register constraint.